syz-manager/manager.go

```diff
@@ -1066,7 +1061,143 @@ func (mgr *Manager) fuzzerConnect() ([]rpctype.RPCInput, BugFrames, bool) {
+		if err != nil {
+			log.Fatalf("failed to update kernel modules: %v", err)
+			return
+		}
+		mgr.modulesInitialized = true
+		mgr.coverFilterBitmap, mgr.coverFilter, err = mgr.updateCoverFilter()
+	})
+	if err != nil {
+		return nil, BugFrames{}, nil, err
+	}
+	r.CoverFilterBitmap = mgr.coverFilterBitmap
+	return corpus, BugFrames{memoryLeaks: memoryLeakFrames, dataRaces: dataRaceFrames}, mgr.coverFilter, nil
+}
+
+const EXT string = ".ko"
+
+func searchModuleName(data []byte) (string, error) {
+	key := []byte("name=")
+	pos := bytes.Index(data, key)
+	if pos == -1 {
+		return "", fmt.Errorf("not find name= keyword")
+	}
+	end := bytes.Index(data[pos:], []byte{0x00})
+	if end == -1 {
+		return "", fmt.Errorf("not find null char")
+	}
+	return string(data[pos+len(key) : pos+end]), nil
+}
+
+func getModuleName(path string) (string, error) {
+	file, err := elf.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open %v", path)
+	}
+	section := file.Section(".modinfo")
+	if section == nil {
+		return "", fmt.Errorf("no .modinfo section")
+	}
+	data, err := section.Data()
+	if err != nil {
+		return "", fmt.Errorf("failed to read .modinfo")
+	}
+	name, err := searchModuleName(data)
+	if err != nil {
+		section = file.Section(".gnu.linkonce.this_module")
+		if section == nil {
+			return "", fmt.Errorf("no .gnu.linkonce.this_module section")
+		}
+		data, err = section.Data()
+		if err != nil {
+			fmt.Println("failed to read .gnu.linkonce.this_module")
+		}
+		name = string(data)
+	}
+	return name, nil
+}
+
+func findModulePaths(dir string) ([]string, error) {
+	files := []string{}
+	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
+		if filepath.Ext(path) == EXT {
+			files = append(files, path)
+		}
+		return nil
+	})
+
+	return files, err
+}
+
+func autoFindModulePath(where, name string) (string, error) {
+	paths, err := findModulePaths(where)
+	if err != nil {
+		return "", err
+	}
+	for _, path := range paths {
+		if strings.TrimSuffix(filepath.Base(path), EXT) == name {
+			return path, nil
+		}
+	}
+	for _, slowpath := range paths {
+		mname, err := getModuleName(slowpath)
+		if err != nil {
+			return "", err
+		}
+		if mname == name {
+			return slowpath, nil
+		}
+	}
+	return "", fmt.Errorf("no match module path found")
+}
+
+func (mgr *Manager) getModulePath(name string) (string, error) {
+	modulePath := ""
+	var err error
+	if mgr.cfg.KernelObj != "" {
+		modulePath, err = autoFindModulePath(mgr.cfg.KernelObj, name)
+	}
+	for _, path := range mgr.cfg.ModuleObj {
+		modulePath, err = autoFindModulePath(path, name)
+		if err == nil {
+			return modulePath, nil
+		}
+	}
+	if err != nil {
+		return "", err
+	}
+	return modulePath, err
+}
+
+func (mgr *Manager) updateKernelModules(a *rpctype.ConnectArgs) error {
+	mgr.KernelModules = make(map[string]backend.KernelModule, len(a.Modules))
+	for _, rmodule := range a.Modules {
+		path, err := mgr.getModulePath(rmodule.Name)
```

Comment:
We call this for every module and it scans the kernel tree. My machine currently has 224, which means we will scan kernel tree 224 times to obtain the same list of modules. This does not look like a good way to structure this logic.
It would be better to scan the tree once to find all modules once, and then do matching in memory. Tree scanning should be done if we have any modules to resolve at all.